### ISSUE TITLE
A practical example (that's not QA) to understand the benefits of DSPy

### ISSUE THREAD

**Issue by ee-lang** [stanfordnlp/dspy#164]
I read the DSP and DSPy papers, but I'm still struggling to understand how to implement certain scenarios or seeing the benefit.

Let's say we have emails that we need to classify (number of classes=N>=5). Based on different classes we need to ask different questions to extract further information. Example: An email is classified as a refund and we'd like to extract the reason for the refund.
The emails can be quite long and contain several email exchanges.

A possible solution to this a problem is to use a zero-shot (encoder only) classifier with hand written labels that covers the semantic meaning of the different classes. Once the text is classified, use another encoder model to do extractive QA. As the emails can be quite long and a lot of unnecessary parts in them, summarization is applied on them before classification.

So the pipeline in a short:
long (email) text -> summary -> classification -> extractive QA (on the original content, based on the class.)

The largest amount of time of this task was spent handcrafting the summarization prompt to capture the relevant details and ignore the rest. After that it, was selecting the right labels for the classification.
Note: our training data was (email, human label) pairs.

My question is, how could DSPy help with such a pipeline?

**Comment by okhat** [comment]
In general, DSPy can help any pipeline through compiling into effective prompts (or finetunes) automatically. We have some examples/* in this repo, besides the intro.ipynb notebook.

Note: We don't usually work with encoder-only models; all LMs work in DSPy through text generation. Even classification can easily be implemented that way (and finetuning encoder-decoder FlanT5 can support this too).

I like your pipeline a lot, and it can be implemented easily in DSPy. One thing I don't understand is why you need the classification before extractive QA. Can you share some intuition about the relation between the two?

**Comment by okhat** [comment]
Or are you saying that classification and QA are unrelated?

**Comment by ee-lang** [comment]
Thanks for the quick response @okhat 

The reason for the classification is that the questions are different:
class=='Refund' -> question:='What is the reason for requesting a refund?'
class=='Invoice/receipt request' -> question:='What is the order number?'
etc.

**Comment by okhat** [comment]
Ah okay, so you already know which classes are needed for classification? And you already know which question to ask for which class? Is that right?

(or do you want this to be automatically discovered?)

**Comment by ee-lang** [comment]
Btw. the putting the classification part aside (I know it can be done with LLMs and the same is true for the extractive QA, but it's not too efficient), what I was curious about is whether DSPy (or any methodology) could help me to find a prompt like in the example, where the search space is really large.

In the current solution the prompt is quite long and contains like 10 lines to pick up the details that are important for the different classes. Is there a way to systematically go about this?

**Comment by okhat** [comment]
Yes definitely, nothing is so hard here in fact. This is a typical task :-)

In terms of efficiency, dspy can compile this into a small t5 model — after prompting works first tho.

The main thing needed is some input emails! You don’t really need many labels. But you need a few emails

**Comment by ee-lang** [comment]
> Ah okay, so you already know which classes are needed for classification? And you already know which question to ask for which class? Is that right?
> 
> (or do you want this to be automatically discovered?)

Yes, I know the classes upfront. But I don't know what would be the best label per class to semantically differentiate them.
The questions are also relatively easy.

Sidenote: I experimented with asking the questions directly and using the perplexity score to figure out whether the class falls into that category at all, but that was problematic as some of the classes are semantically close and some questions "picked up" signal even when the text wasn't about their related class.

**Comment by ee-lang** [comment]
> Yes definitely, nothing is so hard here in fact. This is a typical task :-)
> 
> In terms of efficiency, dspy can compile this into a small t5 model — after prompting works first tho.
> 
> The main thing needed is some input emails! You don’t really need many labels. But you need a few emails

Do you mind helping (or pointing) me to some examples how to do it? :-)

Just for clarify, the hand-crafted prompt ended up being something like this:
"Summarize the text between <EMAIL> tags
Pay attention to:
- if a person is asking a refund
- if a person is asking X
- if a person is asking Y
- ...

<EMAIL>
...
</EMAIL>

Summary:"

**Comment by okhat** [comment]
Yes we can write an example program for this!

but I need to know what you have — do you have example emails? (You don’t need to release them)

**Comment by ee-lang** [comment]
Yes, I have thousands of them. :-)

**Comment by okhat** [comment]
Perfect! This is optional but do you have any labels already for some emails? Even like 3 labels per class is great

**Comment by ee-lang** [comment]
Yes, they're all labeled.

**Comment by ee-lang** [comment]
Worth mentioning though that the emails are often very "messy", lack structure, etc. More importantly they're long, so few-shot-prompting is not really an option.

**Comment by eyast** [comment]
I am in the current XCS224U cohort, and I would love to help with this @okhat

**Comment by okhat** [comment]
Yes @eyast I also would like to do something here. The first thing to figure out is the data, @ee-lang I imagine you cannot share this publicly, right?

Do you know of any similar public datasets? Not necessarily about emails

**Comment by ee-lang** [comment]
I cannot share the real data, but I can write a few examples to simulate
the problem. Give me 1 or 2 days.

**Comment by tomFoxxxx** [comment]
This issue really arouses my interests in DSPy, hope to see more follow-up contents.

**Comment by ee-lang** [comment]
I fabricated 10 examples (5 classes).
https://huggingface.co/datasets/eelang/dsp_examples/blob/main/dsp_ds_10.csv

**Comment by ee-lang** [comment]
The task would be to create a summary with some parts included conditionally (based on the class of the text).

**Comment by eyast** [comment]
I am not affiliated with DSPy, so better wait for Omar's answer.
But I took the liberty of playing around a little bit with the data that you have shared, and I was able to classify the emails with amazing success rate. I'm pushing my code in a gist, just to help you have an idea how to navigate the problem statement from the starting point.. There's so much more that you can achieve with DSPy beyond the (very) simple program here: https://gist.github.com/eyast/acc4b8fdb39fadca9baa2d9307429569#file-email_dspy-py

**Comment by eyast** [comment]
I've had a look at the data and was able to classify using a simple LLM prompt, very easily and without demonstrations. Here's a gist of the code. This represents the first few steps of your pipeline. I believe that based on the reply `pred.category`, you could run other pipelines. If you want, you can treat this problem as a multi-hop search program and embed all the logic in a single class that you can compile with a teleprompted. I personally didn't have to do that.
The next steps - should you want to keep things verbose - would be to right a conditional flow to invoke sub-modules (i.e., if pred == invoice, then do a new prediction to try to extract invoice #").
Here's the code: https://gist.github.com/eyast/acc4b8fdb39fadca9baa2d9307429569#file-email_dspy-py 
I'm not affiliated with DSPy, so i'd love to hear Omar's perspective as well.
Thanks, Eyas

**Comment by eyast** [comment]
@okhat in `dsp`, you could run `dsp.sample` to include demos in an example. is there an equivalent in DSPy?

**Comment by GenAISkills** [comment]
@eyast Many thanks for the example I learnt a lot , Why would one require to include demos in an example can you please eloborate on this Please

**Comment by eyast** [comment]
I come from a background where I have used the previous version of `DSPy` briefly (it was called `DSP`) as part of a classroom course. Effectively speaking, if you do not provide demonstrations, you are simply using LLMs with Zero-shot prompts. By employing demonstrations, you are changing this to a few-shot prompts. The beauty of this library is that you can create long pipelines of many steps (in my example, it was just 1 step), but imagine you have multiple DSP modules 'chained' together. And you need to only provide a demonstration of the first input and last output - anything in between (the in between modules steps) wont require manual labeling (in other words, the middle steps wont require demonstrations). 

The code of DSP is in the DSP folder. I really want to refer to @okhat and the team of the great work they've been doing on these libraries, and I hope they can provide more succinct answers, since I am only learning as well and I want to be mindful not to share mistakes.. I believe that Few-shot prompting in `DSPy` is approached through teleprompters when the programs are compiled, but I have not explore this in full yet, nor have I read the recent DSPy paper in full yet. Sorry if this wasn't too helpful.. still learning :)

**Comment by GenAISkills** [comment]
@eyast  Thanks a lot
I am looking into this library to use cheaper Open source Llama models , would like to know the impact of  under the hood prompt optimizations help , instead of simply hitting the GPT-4 API end point
Which I believe is a good ROI
By the way Can you please  point me to the course you have taken related to DSP
Thanks again

**Comment by eyast** [comment]
for sure! It's a course about Natural Language Understanding and DSP is just a part of it. The course is *amazing*. I believe the lessons are also on Youtube (I believe, please double check). Course info here: https://online.stanford.edu/courses/xcs224u-natural-language-understanding

**Comment by ee-lang** [comment]
> I've had a look at the data and was able to classify using a simple LLM prompt, very easily and without demonstrations. Here's a gist of the code. This represents the first few steps of your pipeline. I believe that based on the reply `pred.category`, you could run other pipelines. If you want, you can treat this problem as a multi-hop search program and embed all the logic in a single class that you can compile with a teleprompted. I personally didn't have to do that. The next steps - should you want to keep things verbose - would be to right a conditional flow to invoke sub-modules (i.e., if pred == invoice, then do a new prediction to try to extract invoice #"). Here's the code: https://gist.github.com/eyast/acc4b8fdb39fadca9baa2d9307429569#file-email_dspy-py I'm not affiliated with DSPy, so i'd love to hear Omar's perspective as well. Thanks, Eyas

Thanks for the example. A couple of observation + extra information:
- You're using gpt-4 which is often not an option commercially. In real world use cases, you're looking at hundreds of thousands or millions of emails per year and expensive models are not a viable option.
- I didn't have the time to write up summaries, but the target really should be that, to create those conditional summaries and extract the relevant information with 95%+ of accuracy.
- As these are fabricated examples, I didn't make them long enough but in real-world use cases you have emails of 1-2000+ tokens. Few shot prompting with a naive approach (i.e. taking the text content as is as a demo) isn't feasible with a 4K context.

**Comment by eyast** [comment]
Hi @ee-lang . You can use whichever GPT model you want that supports completion.
You should preferably provide data that includes two columns: the input text, and the target. if the target is the invoice number for invoice emails, please provide that. You could for example, for the sake of starting quick and easy, just provide invoice numbers, and for non invoice ones, write NaN. Then your pipeline will be a two module program:
1. First module that checks if the email is an invoice email or not
2. A second module that is ran against only emails that return True above, which extracts the invoice.
You don't need to write a classifier for emails. DSPy will do that for your when it 'compiles'.
Please provide the examples to what you want to see. The data you have shared had emails and classifications, that's why I wrote what I wrote.
I hop ethis helps
Cheers,
Eyas

**Comment by eyast** [comment]
hi @ee-lang  - I've experimented a bit more. While zero shot answering seems to be working, I'm unable to compile an program because you haven't provided gold standards. Can I ask you for a favor please, can you add a column to the data you have shared with the final output you expect from the pipeline. For example, the entry that starts with `Dear Acme,  Due to unforeseen circumstances we’d like to request a refun` - what should the final summary be?

To help you ideate a bit better, your data should only contain the first query (the email) and the final output you want (regardless of classification and inner working). The Classification column you've shared is great (for us, learners) but it's not necessarily required to be handcrafted/labeled for each instance. We just need the list of categories to be a list (without necessarily labeling each instance to a class in the training set). 

I also tried gpt3-5-turbo, and got the same results of success (I understand that the data is simple, but I wanted to let you know)

**Comment by ee-lang** [comment]
I'll need a day or two and I'll provide the gold summaries.

**Comment by eyast** [comment]
Thank you. And thanks for having provided the (intermediary) classes. Although they are not necessary for DSPy, they help enhance the understanding and will help validate the work flow of the program.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by okhat** [comment]
Sorry I missed this rich conversation. One key thing: the whole point of DSPy is that you don’t need gold summaries !

it can generate its own summarizes and optimize them to maximize the final metric — please don’t spend your time labeling intermediate steps, DSPy teleprompters do this for you

**Comment by okhat** [comment]
@eyast Thanks a lot for your rich input! Do check out DSPy in isolation, it’s completely different (ie much more powerful and easier to use) compared to the original DSP v1.

**Comment by ee-lang** [comment]
I added the gold summaries.
https://huggingface.co/datasets/eelang/dsp_examples/blob/main/dsp_ds_10.csv

**Comment by ee-lang** [comment]
> Sorry I missed this rich conversation. One key thing: the whole point of DSPy is that you don’t need gold summaries !
> 
> it can generate its own summarizes and optimize them to maximize the final metric — please don’t spend your time labeling intermediate steps, DSPy teleprompters do this for you

The task can be changed to generate a conditional summary (instead of classification). That's a real world scenario that I don't see yet how to solve with DSPy. Can you give us ideas how to approach this?

**Comment by eyast** [comment]
I wrote this script: https://gist.github.com/eyast/994c2731c072727b9d82c439a5c3c169
it does not succeed at creating traces.
I suspect the problem is either with how I chained the signatures in the module, or with my validation logic. I'd love someone else's insight.

**Comment by gabgen** [comment]
Guys, it seems this is the very first practical demo and I have two points : 
1) You started to ask for labels 
2) after 35 comments there is still no solution for a task that you defined "nothing is so hard here in fact. This is a typical task"

I am a supporter of this library but as many already highlighted, it's hard to understand what is the real advantage since the tasks you proposed were relatively simple to do with other libraries. 

Just saying that seeing the final solution, not changing what was the original process but implementing it through the lib, would be very informative of the power of this framework. 

Hope you can provide more on this.

**Comment by ee-lang** [comment]
> Guys, it seems this is the very first practical demo and I have two points :
> 
> 1. You started to ask for labels
> 2. after 35 comments there is still no solution for a task that you defined "nothing is so hard here in fact. This is a typical task"
> 
> I am a supporter of this library but as many already highlighted, it's hard to understand what is the real advantage since the tasks you proposed were relatively simple to do with other libraries. Now, @okhat you would say: "The performance was not great simply asking, means it's hard". Let me clarify: I think what people mean by "hard" is not strictly related to the performance itself. We know that these LLMs already have prior knowledge and already can answer your questions. You can think of a better prompt but the core skill is already there. What people are seeking is a library that easily creates a very specific prompt, tailored to the specific task. In this example, @ee-lang specified that the current solution implies this part : " Pay attention to:
> 
> if a person is asking a refund if a person is asking X if there is an event Y ... "
> 
> the hard part here is not to put good examples, nor to validate it. The hard part is to come up with such a specific prompt only based on the examples. In your intro demo, you only showed something we already knew LLMs are capable of.
> 
> Just saying that seeing the final solution, not changing what was the original process, but implementing it through the lib, that would be very informative of the power of this framework.
> 
> Hope you can provide more on this.

I was able to leverage the philosophy of the DSP framework in tasks that had a clear golden answer (or where you can relatively easily check the answer). Most of the "optimization" is what examples to include in a few-shot scenario (at least in my understanding). In my information extraction use case I ended up re-implementing the parts that I found useful as I didn't see how to add easily my LM of choice (Mistral-Orca) and this seemed the quicker option.

**Comment by gabgen** [comment]
> > Guys, it seems this is the very first practical demo and I have two points :
> > 
> > 1. You started to ask for labels
> > 2. after 35 comments there is still no solution for a task that you defined "nothing is so hard here in fact. This is a typical task"
> > 
> > I am a supporter of this library but as many already highlighted, it's hard to understand what is the real advantage since the tasks you proposed were relatively simple to do with other libraries. Now, @okhat you would say: "The performance was not great simply asking, means it's hard". Let me clarify: I think what people mean by "hard" is not strictly related to the performance itself. We know that these LLMs already have prior knowledge and already can answer your questions. You can think of a better prompt but the core skill is already there. What people are seeking is a library that easily creates a very specific prompt, tailored to the specific task. In this example, @ee-lang specified that the current solution implies this part : " Pay attention to:
> > if a person is asking a refund if a person is asking X if there is an event Y ... "
> > the hard part here is not to put good examples, nor to validate it. The hard part is to come up with such a specific prompt only based on the examples. In your intro demo, you only showed something we already knew LLMs are capable of.
> > Just saying that seeing the final solution, not changing what was the original process, but implementing it through the lib, that would be very informative of the power of this framework.
> > Hope you can provide more on this.
> 
> I was able to leverage the philosophy of the DSP framework in tasks that had a clear golden answer (or where you can relatively easily check the answer). Most of the "optimization" is what examples to include in a few-shot scenario (at least in my understanding). In my information extraction use case I ended up re-implementing the parts that I found useful as I didn't see how to add easily my LM of choice (Mistral-Orca) and this seemed the quicker option.

Could you share your solution?

**Comment by okhat** [comment]
Happy to see lots of engagement here folks!

I maintain this is a simple task, well within the capacity of the framework. People have built at least a dozen non-factoid programs in DSPy that I know of, probably many more! It would be awesome if we could build an example here for this, but we’re busy adding new features.

That said, I’ll see if I can get a student on this task.

**Comment by okhat** [comment]
I think the first non-factoid example built with this library was in November 2022, more than a year ago. Expect more releases soon :-)

**Comment by darinkishore** [comment]
There's one I created on the discord examples section:

https://discord.gg/gb5k34m4pm

**Comment by ahoho** [comment]
I agree---having a simple retrieval-based ICL classification example would be very helpful (especially if it didn't involve a ColBERT installation).

**Comment by okhat** [comment]
There’s now a whole bunch of non-QA examples in examples/ and throughout other places on Discord and Twitter — there are still some cool threads in this issue but closing until someone takes a look at playing with them

**Comment by zendesk-ryan-li** [comment]
@darinkishore  wonder where can I check the example? the link is no longer valid
